Public/Src/Tools/SymbolDaemon/VsoSymbolClient.cs

```diff
@@ -0,0 +1,208 @@
+            }
+        }
+
+        /// <nodoc />
+        public VsoSymbolClient(ILogger logger, SymbolConfig config)
+        {
+            m_logger = logger;
+            m_config = config;
+            m_cancellationSource = new CancellationTokenSource();
+
+            m_logger.Info(I($"[{nameof(VsoSymbolClient)}] Using drop config: {JsonConvert.SerializeObject(m_config)}"));
+
+            m_symbolClient = CreateSymbolServiceClient();
+        }
+
+        private ISymbolServiceClient CreateSymbolServiceClient()
+        {
+            var client = new SymbolServiceClient(
+                ServiceEndpoint,
+                GetFactory(),
+                Tracer,
+                new SymbolServiceClientTelemetry(Tracer, ServiceEndpoint, enable: m_config.EnableTelemetry));
+
+            return client;
+        }
+
+        /// <summary>
+        /// Queries the endpoint for the RequestId. 
+        /// This method should be called only after the request has been created, otherwise, it will throw an exception.
+        /// </summary>
+        /// <remarks>
+        /// On workers, m_requestId won't be initialized, so we need to query the server for the right value.
+        /// </remarks>
+        private async Task EnsureRequestIdInitalizedAsync()
+        {
+            if (string.IsNullOrEmpty(m_requestId))
+            {
+                var result = await m_symbolClient.GetRequestByNameAsync(RequestName, Token);
+                m_requestId = result.Id;
+            }
+        }
+
+        /// <summary>
+        /// Creates a symbol request.
+        /// </summary>
+        public async Task<Request> CreateAsync(CancellationToken token)
+        {
+            var result = await m_symbolClient.CreateRequestAsync(RequestName, token);
+
+            m_requestId = result.Id;
+
+            return result;
+        }
+
+        /// <inheritdoc />
+        public Task<Request> CreateAsync() => CreateAsync(Token);
+
+        /// <inheritdoc />
+        public async Task<AddDebugEntryResult> AddFileAsync(SymbolFile symbolFile)
+        {
+            Contract.Requires(symbolFile.IsIndexed, "File has not been indexed.");
+            Contract.Requires(symbolFile.DebugEntries.Count > 0, "File contains no symbol data.");
+
+            await EnsureRequestIdInitalizedAsync();
+
+            var result = await m_symbolClient.CreateRequestDebugEntriesAsync(
+                RequestId,
+                symbolFile.DebugEntries.Select(e => CreateDebugEntry(e)),
+                DefaultDebugEntryCreateBehavior,
+                Token);
+
+            var entriesWithMissingBlobs = result.Where(e => e.Status == DebugEntryStatus.BlobMissing).ToList();
+
+            if (entriesWithMissingBlobs.Count > 0)
+            {
+                // All the entries are based on the same file, so we need to call upload only once.
+
+                // make sure that the file is on disk
+                var file = await symbolFile.EnsureMaterializedAsync();
+

```

Comment:
This is called twice per file because we materialize the file before calling AddFileAsync. 

symbolFile.IsIndexed is enough to assume that the file is already materialized, right?

Reply:
Currently - yes. Once I add caching, we'll be calling into materialize only here.

---
In reply to: [314018861](https://github.com/microsoft/BuildXL/pull/727#discussion_r314018861) [](ancestors = 314018861)